File: tests/mathematical/pi_calculation/pi_calculation.py
#!/usr/bin/env python3
"""
Pi calculation benchmark implementation in Python.
Calculates pi using the Monte Carlo method.
"""

import sys
import time
import random
import math
import numpy as np


def calculate_pi_monte_carlo(num_samples):
    """Calculate pi using Monte Carlo method with NumPy for vectorization."""
    # Generate random coordinates in a vectorized manner
    x = np.random.rand(num_samples)
    y = np.random.rand(num_samples)
    
    # Use vectorized operations to find points inside the circle
    inside_circle = np.sum(x*x + y*y <= 1)
    
    return 4.0 * inside_circle / num_samples


def main():
    """Main execution function."""
    num_samples = 1000000
    
    print(f"Calculating pi with {num_samples} samples...")
    start_time = time.perf_counter()
    
    pi_estimate = calculate_pi_monte_carlo(num_samples)
    
    end_time = time.perf_counter()
    execution_time = end_time - start_time
    
    error = abs(pi_estimate - math.pi)
    
    print(f"Result: {pi_estimate:.6f}")
    print(f"Actual pi: {math.pi:.6f}")
    print(f"Error: {error:.6f}")
    print(f"Execution time: {execution_time:.6f} seconds")


if __name__ == "__main__":
    main()

File: tests/mathematical/pi_calculation/pi_calculation.ts
#!/usr/bin/env node
/**
 * Pi calculation benchmark implementation in TypeScript.
 * Calculates pi using the Monte Carlo method.
 */

function calculatePiMonteCarlo(numSamples: number): number {
    let insideCircle = 0;
    
    for (let i = 0; i < numSamples; i++) {
        const x = Math.random();
        const y = Math.random();
        
        if (x*x + y*y <= 1) {
            insideCircle++;
        }
    }
    
    return 4.0 * insideCircle / numSamples;
}

function main(): void {
    const numSamples = 1000000;
    
    console.log(`Calculating pi with ${numSamples} samples...`);
    const start = performance.now();
    
    const piEstimate = calculatePiMonteCarlo(numSamples);
    
    const end = performance.now();
    const executionTime = (end - start) / 1000; // Convert to seconds
    
    const actualPi = Math.PI;
    const error = Math.abs(piEstimate - actualPi);
    
    console.log(`Result: ${piEstimate.toFixed(6)}`);
    console.log(`Actual pi: ${actualPi.toFixed(6)}`);
    console.log(`Error: ${error.toFixed(6)}`);
    console.log(`Execution time: ${executionTime.toFixed(6)} seconds`);
}

if (require.main === module) {
    main();
}